src/poetry/console/commands/self/self_command.py

```diff
@@ -0,0 +1,119 @@
+
+
+class SelfCommand(InstallerCommand):
+    ADDITIONAL_PACKAGE_GROUP = "additional"
+
+    @staticmethod
+    def get_default_system_pyproject_file() -> Path:
+        # We separate this out to avoid unwanted side effect during testing while
+        # maintaining dynamic use in help text.
+        #
+        # This is not ideal, but is the simplest solution for now.
+        from poetry.locations import CONFIG_DIR
+
+        return Path(CONFIG_DIR).joinpath("pyproject.toml")
+
+    @property
+    def system_pyproject(self) -> Path:
+        file = self.get_default_system_pyproject_file()
+        file.parent.mkdir(parents=True, exist_ok=True)
+        return file
+
+    def reset_env(self) -> None:
+        self._env = EnvManager.get_system_env(naive=True)
+
+    @property
+    def env(self) -> Env:
+        if self._env is None or not isinstance(self._env, SystemEnv):
+            self.reset_env()
+        return self._env
+
+    @property
+    def default_group(self) -> str:
+        return self.ADDITIONAL_PACKAGE_GROUP
+
+    @property
+    def activated_groups(self) -> set[str]:
+        return {self.default_group}
+
+    @contextmanager
+    def with_system_project(self) -> Iterator[Path]:
+        cwd = Path.cwd()
+        try:
+            os.chdir(self.system_pyproject.parent)
+            yield self.system_pyproject.parent
+        finally:
+            os.chdir(cwd)
+
+    def generate_system_pyproject(self) -> None:
+        preserved = {}
+
+        if self.system_pyproject.exists():
+            content = PyProjectTOML(self.system_pyproject).poetry_config
+
+            for key in {"group", "source"}:
+                if key in content:
+                    preserved[key] = content[key]
+
+        package = ProjectPackage(name="poetry-instance", version=__version__)
+        package.add_dependency(Dependency(name="poetry", constraint=__version__))
+
+        package.python_versions = ".".join(str(v) for v in self.env.version_info[:3])
+
+        content = Factory.create_pyproject_from_package(package=package)
+
+        for key in preserved:
+            content[key] = preserved[key]
+
+        self.system_pyproject.write_text(content.as_string(), encoding="utf-8")
+
+    def reset_poetry(self) -> None:
+        with self.with_system_project():
+            self.generate_system_pyproject()
+            self._poetry = Factory().create_poetry(
+                self.system_pyproject.parent, io=self._io, disable_plugins=True
+            )
+
+    @property
+    def poetry(self) -> Poetry:
+        if self._poetry is None:
+            self.reset_poetry()
+        assert self._poetry
+        return self._poetry
+
+    def _system_project_handle(self) -> int:
+        return super().handle()
+
+    def reset(self) -> None:
+        self.reset_env()
+        self.reset_poetry()
+
+    def handle(self) -> int:
+        self.reset()
+
+        with self.with_system_project():
+            return self._system_project_handle()
```

Comment:
Can you please explain when and how this method is used?

The other `self` commands inherit from `SelfCommand` and doesn't overwrite the `handle` method. In case of classes that inherit from multiple base classes `SelfCommand` is always the most left one, so it's implementation would take precedence over implementation in the class(es), wouldn't it?

Reply:
If I am understanding your concern correctly, the short answer is [C3 linearlization](https://en.wikipedia.org/wiki/C3_linearization) ensures the order of execution is defined by the [MRO](https://www.python.org/download/releases/2.3/mro/) ([simplified](https://www.educative.io/edpresso/what-is-mro-in-python)) in the child class not parent/siblings. The example below should show a cut down version of what is happening.

<details>
<summary><b>example.py</b></summary>
<p>

```py
from __future__ import annotations


def mro(clz: type) -> None:
    print(f"MRO({clz.__name__}): ", " -> ".join(map(lambda m: m.__name__, clz.mro())))


class InstallerCommand:
    def handle(self) -> None:
        print("InstallerCommand.handle() from ", self.__class__.__name__)
        mro(InstallerCommand)
        raise NotImplementedError()


class InstallCommand(InstallerCommand):
    def handle(self) -> None:
        print("InstallCommand.handle() from ", self.__class__.__name__)
        mro(InstallCommand)


class SelfCommand(InstallerCommand):
    def _system_project_handle(self) -> None:
        print("SelfCommand._system_project_handle() from ", self.__class__.__name__)
        super().handle()

    def handle(self) -> None:
        print("SelfCommand.handle() from ", self.__class__.__name__)
        mro(SelfCommand)
        self._system_project_handle()


class SelfInstallCommand(SelfCommand, InstallCommand):
    # this is a passthrough command, ie. it modifies options and description only
    pass


class SelfCustomCommand(SelfCommand):
    # this command is a new command that is not wrapping an existing command, or needs
    # pre-processing prior to calling another command

    def _system_project_handle(self) -> None:
        print(
            "SelfCustomCommand._system_project_handle() from ", self.__class__.__name__
        )
        mro(SelfCustomCommand)


if __name__ == "__main__":
    print("##########################")
    mro(SelfInstallCommand)
    SelfInstallCommand().handle()
    print("##########################")
    mro(SelfCustomCommand)
    SelfCustomCommand().handle()
    print("##########################")

```

</p>
</details>

The above code will output something like this.
```console
##########################
MRO(SelfInstallCommand):  SelfInstallCommand -> SelfCommand -> InstallCommand -> InstallerCommand -> object
SelfCommand.handle() from  SelfInstallCommand
MRO(SelfCommand):  SelfCommand -> InstallerCommand -> object
SelfCommand._system_project_handle() from  SelfInstallCommand
InstallCommand.handle() from  SelfInstallCommand
MRO(InstallCommand):  InstallCommand -> InstallerCommand -> object
##########################
MRO(SelfCustomCommand):  SelfCustomCommand -> SelfCommand -> InstallerCommand -> object
SelfCommand.handle() from  SelfCustomCommand
MRO(SelfCommand):  SelfCommand -> InstallerCommand -> object
SelfCustomCommand._system_project_handle() from  SelfCustomCommand
MRO(SelfCustomCommand):  SelfCustomCommand -> SelfCommand -> InstallerCommand -> object
##########################
```

Regarding when it is used, self commands are currently designed for 2 use cases.

1. A wrapper to poetry top level commands like (add, remove, show) that only differs in that it manages the "poetry system" project; and may have different options/arguments/descriptions etc.
2. A command that either has its own handle logic or requiring pre-processing prior to executing the system project.

The `SelfCommand` base class handles the boilerplate of switching context and resetting the instance. The `_system_project_handle` method acts as an in-context overwritable hook for child classes.

I will add some additional in-code documentation later.

Reply:
I was right with the interpretation of what `handle` method is called in case of multiple inheritance. I just overlooked that `_system_project_handle` actually calls the `handle` method of the "non-self command". Got it now :+1: